src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -4166,16 +4167,208 @@ private void ImportCompareOperation(ILOpcode opcode)
 
             // Load the value and then convert it instead of using ValueAsType to avoid loading the incorrect size
             LLVMValueRef loadedValue = value.ValueAsType(value.Type, _builder);
+
+            if (checkOverflow)
+            {
+                Debug.Assert(destType is EcmaType);
+                BuildConvOverflow(loadedValue, (EcmaType)destType, wellKnownType, unsigned, value.Type);
+            }
+
             LLVMValueRef converted = CastIfNecessary(loadedValue, GetLLVMTypeForTypeDesc(destType), value.Name(), wellKnownType == WellKnownType.UInt64 /* unsigned is always false, so check for the type explicitly */);
             PushExpression(GetStackValueKind(destType), "conv", converted, destType);
         }
 
+        private bool IsLlvmReal(LLVMTypeRef llvmTypeRef)
+        {
+            return llvmTypeRef == LLVMTypeRef.Float || llvmTypeRef == LLVMTypeRef.Double;
+        }
+
+        void BuildConvOverflow(LLVMValueRef loadedValue, EcmaType destType, WellKnownType destWellKnownType, bool unsigned, TypeDesc sourceType)
+        {
+            ulong maxValue = 0;
+            long minValue = 0;
+            switch (destWellKnownType)
+            {
+                case WellKnownType.Byte:
+                    maxValue = byte.MaxValue;
+                    break;
+                case WellKnownType.SByte:
+                    maxValue = (ulong)sbyte.MaxValue;
+                    minValue = sbyte.MinValue;
+                    break;
+                case WellKnownType.UInt16:
+                    maxValue = ushort.MaxValue;
+                    break;
+                case WellKnownType.Int16:
+                    maxValue = (ulong)short.MaxValue;
+                    minValue = short.MinValue;
+                    break;
+                case WellKnownType.UInt32:
+                case WellKnownType.UIntPtr: // TODO : 64bit.
+                    maxValue = uint.MaxValue;
+                    break;
+                case WellKnownType.Int32:
+                    maxValue = int.MaxValue;
+                    minValue = int.MinValue;
+                    break;
+                case WellKnownType.UInt64:
+                    maxValue = ulong.MaxValue;
+                    break;
+                case WellKnownType.Int64:
+                    maxValue = long.MaxValue;
+                    minValue = long.MinValue;
+                    break;
+            }
+            if (unsigned)
+            {
+                BuildPositiveConvChecks(loadedValue, unsigned, maxValue, sourceType, destType);
+            }
+            else
+            {
+                // for signed overflow, first check for postive or negative and then test appropriately.
+                LLVMValueRef posTest;
+                if (IsLlvmReal(loadedValue.TypeOf))
+                {
+                    posTest = _builder.BuildFCmp(LLVMRealPredicate.LLVMRealUGE, loadedValue, LLVMValueRef.CreateConstReal(loadedValue.TypeOf, 0)); // TODO: NaN/Infinity?
```

Comment:
It may be better to call `Dbl2IntOvf` and similar helpers from `Internal.Runtime.CompilerHelpers.MathHelpers`. They handle NaNs and other corner cases correctly.

Reply:
Doh, didn't see those.  Will replace my with those calls.

Reply:
I'm a little confused how these helpers deal with i8 and i16 types.  I understand these types and the result are expanded to i32 on evaluation stack, but the overflow check still needs to occur for the target type.  Are these small types handled differently?

Reply:
RyuJIT treats them as two operations: convert float to Int32 first and then Int32 to the small type second.

Reply:
That sounds like 2 overflow checks, as I've already got code for i16 and i8 types, is it ok if I leave them as is and use `MathHelpers` for the i32/64 types only?

Reply:
Most of this makes sense, for float -> ulong, I cant use the float helper methods are there are only `Flt2IntOvf` and `Flt2LngOvf` neither of which would handle `ulong - 1` (maybe that can't be represented in a float, but whatever would go in), so for float -> ulong, do I first convert to a double so I can use `Dbl2ULngOvf`?

Reply:
You can do that, or you can add a few more helpers for this. Either way is fine.

Reply:
Thanks, have removed a lot of the logic and simplified I hope to address the feedback.  I went with just extending floats to doubles in the end.